content/blog/hard-grids-and-baselines-android-design-fidelity/index.md

```diff
@@ -0,0 +1,289 @@
+
+When you use a TextView, it has one parameter turned on by default: **includeFontPadding**. includeFontPadding increases the height of a TextView to give room to ascenders and descenders that might not fit within the regular bounds.
+
+![A comparison between having "includeFontPadding" on and off. When it's off the height is 19sp and when it's on it is 21.33sp. It shows the formula "includeFontPadding = TextSize * 1.33"](./images/includeFontPadding.png "A comparison of having the 'includeFontPadding' property enabled")
+
+Now that we know how Android’s typography works, let’s look at an example.
+
+Here’s a simple mockup, detailing the spacing between a title and a subtitle. It is built at 1x, with Figma, meaning line height defines the final height of a text box — not the text size. (This is how most design tools work)
+
+![A spec file of a phone dailing application](./images/Specs.png)
+
+![A mockup with spec lines enabled of a call log app](./images/Implementation.png )
+
+*Of course, because it’s Android, the line height has no effect on the height of the TextView, and the layout is therefore 8dp too short of the mockups.*
+
+But even if it did have an effect, the problems wouldn’t stop there; the issue is more complex than that.
+
+## What designers want, and what developers can do
+
+Designers, like myself, like to see perfect alignment. We like consistent values and visual rhythm.
+
+![A showcase of the differences in line spacing between a mockup and an implementation in Android](./images/Designers_Want_Designers_Get.png)
+
+Unfortunately, translating values from a design tool wasn’t possible. You had the option to either pixel nudge (pictured above, right), or forget about alignment altogether thus leading to an incorrect implementation that would, yet again, be shorter than the mockups.
+
+### …Until now!
+
+*firstBaselineToTopHeight* and *lastBaselineToBottomHeight* are powerful tools for Android design. They do as the name suggests: If *firstBaselineToTopHeight *is set to 56sp, then that’ll become the distance between the first baseline and the top of a TextView.
+
+![A subtitle block showing 56sp height despite the text visually being much shorter](./images/56sp.png)
+
+This means that designers, alongside developers, can force the bounds of a TextView to match the design specs and open the door to perfect implementations of their mockups.
+
+This is something I’ve personally tested in an app I designed. [**Memoire**, a note taking app](http://tiny.cc/getmemoire) for Android, is a 1:1 recreation of its mockups — for every single screen. This was made possible due to these APIs — *and because [**@sasikanth**](https://twitter.com/its\_sasikanth) is not confrontational* — , since text is what almost always makes baseline alignment and hard grids impossible to implement in production.
+
+![Near-perfect duplication of guidelines against Memoire's mockups and actual app](./images/Memoire_Bounds_and_Baselines.gif)
```

Comment:
This image has the magnification cursor applied, but it doesn't become a Gatsby zoome image. Why is this?

Reply:
The only thing that I can see that would seem to distinguish that image from the other GIFs is that it's slightly over 1 MB. Could that be the issue? Considering that the majority of browsers these days support it, it might be a decent idea to try using APNG when possible with an `srcset`. I'm not quite sure if that would be possible in markdown without things getting ugly but it's an idea at least.

Reply:
Rather than starting to use APNG, I would like to see `<video>` (with webm support) being used since it has built-in magnification and would solve GIFs being a [Pause, Stop, Hide](https://www.w3.org/WAI/WCAG21/Understanding/pause-stop-hide.html) violation.

Reply:
More thoughts from PMs with @adueppen:

- Using `aria-described-by` or `title` to the video tag
- Something like the [Material Design guidelines](https://material.io/design/motion/understanding-motion.html#transition-anatomy) player
	- Dark overlay and play button shown by default
	- Autoplay only when needed or allow an option to turn it off for the entire site 
	- Auto pause when left the viewport to reduce jank on mobile

Reply:
I have now changed away from using GIFs into HTML5 videos. We have more work that can be done, but AFAIK the user can change their browser settings to handle the videos a particular way so we should be good for an initial release